ndcube/ndcube.py

```diff
@@ -362,6 +362,93 @@ def axis_world_coords(self, *axes, edges=False):
         else:
             return tuple(axes_coords)
 
+    def axis_world_coords_values(self, *axes, edges=False):
+        """
+        Returns WCS coordinate values of all pixels for desired axes.
+
+        Parameters
+        ----------
+        axes: `int` or `str`, or multiple `int` or `str`
+            Axis number in numpy ordering or unique substring of
+            `~ndcube.NDCube.wcs.world_axis_physical_types`
+            of axes for which real world coordinates are desired.
+            axes=None implies all axes will be returned.
+
+        edges: `bool`
+            If True, the coords at the edges of the pixels are returned
+            rather than the coords at the center of the pixels.
+            Note that there are n+1 edges for n pixels which is reflected
+            in the returned coords.
+            Default=False, i.e. pixel centers are returned.
+
+        Returns
+        -------
+        axis_names: `tuple` of `str`
+            The physical types of the coords returned.
+
+        axes_coords: `tuple` of `astropy.units.Quantity`
+            Real world coords for axes requested by user.
+            Returned in same order as axis_names.
+
+        Example
+        -------
+        >>> NDCube.all_world_coords_values(('lat', 'lon')) # doctest: +SKIP
+        >>> NDCube.all_world_coords_values(2) # doctest: +SKIP
+
+        """
+        # Create meshgrid of all pixel coordinates.
+        # If user, wants edges, set pixel values to pixel edges.
+        # Else make pixel centers.
+        wcs_shape = self.data.shape[::-1]
+        if edges:
+            wcs_shape = tuple(np.array(wcs_shape) + 1)
+            pixel_inputs = np.meshgrid(*[np.arange(i) - 0.5 for i in wcs_shape], indexing='ij')
+        else:
+            pixel_inputs = np.meshgrid(*[np.arange(i) for i in wcs_shape], indexing='ij')
```

Comment:
This needs some thought, at the moment I am pretty sure this is memory dense.

Reply:
OK.  It could well be.  I didn't really understand how it could not be.  If you have a way to ensure memory efficiency is maximized that would be great.  But would it be a case of changing this one line?  Or would the whole approach need to be refactored?

Reply:
I compared the outputs of ```np.meshgrid``` using the ```sparse``` kwargs.  For ```wcs_shape=(32, 1024, 48, 1)``` and the same number of pixel and world dimensions, I got the following sizes in bytes:
| ```sparse``` | Meshgrid List |  Meshgrid[0] |  Meshgrid[1] |  Meshgrid[2] |  Meshgrid[3] |
|--------------|:--------------:|-------------:|-------------:|-------------:|--------------:|
| ```False```   |  104                  | 12583056     | 12583056    | 12583056     | 12583056      |
| ```True```    |   104                 | 400                | 8336            | 528                | 152                 |

Meanwhile the sizes for the outputs of ```pixel_to_world``` was the same independent of whether ```sparse``` was ```True``` or ```False```.  The checked the sizes of the output ```Quantities``` and the ```Quantity.values```.  They were:
|                 | coord0 | coord1 | coord2 | coord3 |
|----------|:-------:|-------:|--------:|-------:|
| Quantity  | 176       | 176      | 176       | 64       |
| Values     | 144       | 144     | 144       | 12583056|

For clarity, these were the sizes before redundant dimensions were removed.  And coord3 was a ```Time``` object which I imagine explains the much large size of its ```values```.

Is this sufficient to assuage memory concerns or is there are way to make the inputs or outputs sparser?  Or is the concern that even if the input is sparse, it must be made dense as part of the calculations and so this doesn't save us much?

An alternative approach would be to perform the transformations separately for sets of dependent axes.  In the above example, axes ```0```, ```1```, and ```3``` are dependent, so we would call ```pixel_to_world``` twice, once for a pixel grid of size ```(32, 1024, 1, 1)``` and once for a grid of ```(1, 1, 48, 1)```.  However, this seems like it would reintroduce the code complexity of the old version of ```axes_world_coords```.  Since as well as sorting out the user input, breaking up the pixel grid was the main source of code complexity.